File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/concurrent/InnerPolyline.cpp
/*
 * InnerPolyline.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <utility>
#include <vector>
#include "InnerPolyline.h"

InnerPolyline::InnerPolyline(std::vector<Point> raw_coordinates, Shape* shape)
  : raw_coordinates_(std::move(raw_coordinates)),
    shape(shape) {}
InnerPolyline::InnerPolyline(Sequence* sequence) {
  this->raw_coordinates_ = sequence->to_vector();
  this->shape = sequence->shape;
}

void InnerPolyline::compute_vertical_bounds()
{ if (!raw_coordinates_.empty()) {
    int min_y = raw_coordinates_[0].y;
    int max_y = raw_coordinates_[0].y;
    for (const auto& pos : raw_coordinates_) {
      if (pos.y < min_y) min_y = pos.y;
      if (pos.y > max_y) max_y = pos.y;
    }
    vertical_bounds_.min = min_y;
    vertical_bounds_.max = max_y;
  }
}

File: lib/contrek/finder/concurrent/inner_polyline.rb
# frozen_string_literal: true

module Contrek
  module Concurrent
    class InnerPolyline
      attr_reader :raw, :vertical_bounds
      attr_accessor :shape

      def initialize(shape: nil, raw_coordinates: [], sequence: nil)
        if sequence
          @shape = sequence.shape
          @raw = sequence.to_a
        else
          @shape = shape
          @raw = raw_coordinates
        end
      end

      def compute_vertical_bounds!
        @vertical_bounds ||= begin
          min_y = Float::INFINITY
          max_y = 0
          @raw.each do |pos|
            y = pos[:y]
            min_y = y if y < min_y
            max_y = y if y > max_y
          end
          {min: min_y, max: max_y}
        end
      end
    end
  end
end
